src/Build/BackEnd/Components/ResourceManager/ResourceManagerService.cs

```diff
@@ -0,0 +1,114 @@
+        {
+            ErrorUtilities.VerifyThrow(type == BuildComponentType.TaskResourceManager, "Cannot create components of type {0}", type);
+
+            return new ResourceManagerService();
+        }
+
+        public void InitializeComponent(IBuildComponentHost host)
+        {
+            string semaphoreName = host.BuildParameters.ResourceManagerSemaphoreName;
+
+            int resourceCount = host.BuildParameters.MaxNodeCount; // TODO: tweakability
+
+#if DEBUG
+            TotalNumberHeld = 0;
+            SemaphoreName = semaphoreName;
+#endif
+
+            if (NativeMethodsShared.IsWindows)
+            {
+                s = new Semaphore(resourceCount, resourceCount, semaphoreName); // TODO: SemaphoreSecurity?
+            }
+            else
+            {
+                // UNDONE: just don't support gathering additional cores on non-Windows
+                s = new Semaphore(1, 1);
+            }
+        }
+
+        public void ShutdownComponent()
+        {
+            s?.Dispose();
+            s = null;
+
+#if DEBUG
+            TotalNumberHeld = -2;
+#endif
+        }
+
+        public int RequestCores(int requestedCores)
+        {
+            if (s is null)
+            {
+                if (!NativeMethodsShared.IsWindows)
+                {
+                    // Since the current implementation of the cross-process resource count uses
+                    // named semaphores, it's not usable on non-Windows, so just return the
+                    // guaranteed resource.
+                    return 1;
+                }
+
+                throw new InternalErrorException($"{nameof(ResourceManagerService)} was called while uninitialized");
+            }
+
+            int i = 0;
+
+            // First core gets a blocking wait: the user task wants to do *something*
+            s.WaitOne();
+
+            // Keep requesting cores until we can't anymore, or we've gotten the number of cores we wanted.
+            for (i = 1; i < requestedCores; i++)
+            {
+                if (!s.WaitOne(0))
+                {
+                    return i;
+                }
+            }
+
+            return i;
+        }
+
+        public void ReleaseCores(int coresToRelease)
+        {
+            if (s is null)
+            {
+                if (!NativeMethodsShared.IsWindows)
+                {
+                    // Since the current implementation of the cross-process resource count uses
+                    // named semaphores, it's not usable on non-Windows, so just continue.
+                    return;
+                }
+
+                throw new InternalErrorException($"{nameof(ResourceManagerService)} was called while uninitialized");
+            }
+
+            ErrorUtilities.VerifyThrow(coresToRelease > 0, "Tried to release {0} cores", coresToRelease);
+
+            s.Release(coresToRelease);
+
+#if DEBUG
+            TotalNumberHeld -= coresToRelease;
```

Comment:
How is TotalNumberHeld actually used? It looks like you just reset it.

Reply:
Can we add some diagnostic logging to be able to see how the cores are requested/released?

Reply:
ETW traces?